compiler/src/iree/compiler/Codegen/Common/MaterializeTuningSpecsPass.cpp

```diff
@@ -0,0 +1,166 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <cassert>
+#include "iree/compiler/Codegen/Common/Passes.h"
+#include "iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenAttrs.h"
+#include "iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenDialect.h"
+#include "llvm/ADT/SmallString.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/ToolOutputFile.h"
+#include "llvm/Support/raw_ostream.h"
+#include "mlir/Bytecode/BytecodeWriter.h"
+#include "mlir/Dialect/Transform/IR/TransformDialect.h"
+#include "mlir/Dialect/Transform/IR/TransformOps.h"
+#include "mlir/Dialect/Transform/IR/TransformTypes.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/BuiltinTypeInterfaces.h"
+#include "mlir/IR/Location.h"
+#include "mlir/IR/OwningOpRef.h"
+#include "mlir/Support/FileUtilities.h"
+
+#define DEBUG_TYPE "iree-codegen-materialize-tuning-specs"
+#define DBGS() (llvm::dbgs() << "[" DEBUG_TYPE "]: ")
+#define LDBG(X) LLVM_DEBUG(DBGS() << X << "\n")
+
+namespace mlir::iree_compiler {
+
+#define GEN_PASS_DEF_MATERIALIZETUNINGSPECSPASS
+#include "iree/compiler/Codegen/Common/Passes.h.inc"
+
+namespace {
+
+llvm::cl::opt<std::string> clCodegenTuningSpecPath(
+    "iree-codegen-tuning-spec-path",
+    llvm::cl::desc("File path to a module containing a tuning spec (transform "
+                   "dialect library)."),
+    llvm::cl::init(""));
+
+llvm::cl::opt<std::string> clCodegenTuningSpecDumpDir(
+    "iree-codegen-dump-tuning-specs-to",
+    llvm::cl::desc(
+        "Dump the final tuning spec modules to the specified directory. When "
+        "set to '-', prints the tuning spec to stdout."),
+    llvm::cl::init(""));
+
+using mlir::transform::NamedSequenceOp;
+
+static LogicalResult dumpFinalTuningSpecToDir(ModuleOp tuningSpec,
+                                              StringRef dir) {
+  if (dir == "-") {
+    tuningSpec->print(llvm::outs());
+    return success();
+  }
+
+  llvm::sys::fs::create_directories(dir);
```

Comment:
Should check what happens on Windows. Last time I used llvm::sys::fs it caused issues on Windows.

Reply:
I copied this code from the `--iree-hal-dump-executable-files-to` logic, so I'd think it should work in both places, although I haven't checked on windows. If we learn it doesn't work, I can fix it.
```console
➜  ag fs::create_directories                  
src/iree/compiler/Codegen/Common/MaterializeTuningSpecsPass.cpp
60:  llvm::sys::fs::create_directories(dir);

src/iree/compiler/Dialect/Util/Transforms/DumpModule.cpp
29:    llvm::sys::fs::create_directories(llvm::sys::path::parent_path(path));

src/iree/compiler/Dialect/HAL/Transforms/SerializeExecutables.cpp
72:      llvm::sys::fs::create_directories(dumpIntermediatesPath);
75:      llvm::sys::fs::create_directories(dumpBinariesPath);

src/iree/compiler/Dialect/HAL/Transforms/DumpExecutableSources.cpp
50:      llvm::sys::fs::create_directories(path);

src/iree/compiler/Dialect/HAL/Transforms/DumpExecutableBenchmarks.cpp
514:      llvm::sys::fs::create_directories(path);
```